lib/mongo/grid_fs/bucket.ex

```diff
@@ -0,0 +1,229 @@
+
+  """
+
+  alias BSON.ObjectId
+  alias Mongo.GridFs.Bucket
+
+  ##
+  # constants used in this module
+  #
+  @files_index_name "filename_1_uploadDate_1"
+  @chunks_index_name "files_id_1_n_1"
+  @defaults [name: "fs", chunk_size: 255 * 1024]
+
+  @type t :: %__MODULE__{
+               name: String.t,
+               chunk_size: non_neg_integer,
+               topology_pid: GenServer.server
+             }
+
+  defstruct name: "fs", chunk_size: 255 * 1024, topology_pid: nil, opts: []
+
+  @doc """
+  Creates a new Bucket with a existing connection using the default values. It just contains the
+  name of the collections (fs) and the chunk size (255KB).
+
+  The bucket checks the index for both collections as well. In case of multiple
+  upload or downloads just create only one bucket and reuse it.
+
+  """
+  @spec new(GenServer.server, Keyword.t) :: Bucket.t
+  def new(topology_pid, options \\ []) do
+
+    Keyword.merge(@defaults, options)
+    |> Enum.reduce(%Bucket{topology_pid: topology_pid, opts: options}, fn {k, v}, bucket -> Map.put(bucket, k, v) end)
+    |> check_indexes
+
+  end
+
+  @doc """
+  Returns the collection name for the files collection, default is fs.files.
+  """
+  @spec files_collection_name(Bucket.t) :: String.t
+  def files_collection_name(%Bucket{name: fs}), do: "#{fs}.files"
+
+  @doc """
+  Returns the collection name for the chunks collection, default is fs.chunks.
+  """
+  @spec chunks_collection_name(Bucket.t) :: String.t
+  def chunks_collection_name(%Bucket{name: fs}), do: "#{fs}.chunks"
+
+  @doc """
+  Renames the stored file with the specified file_id.
+  """
+  @spec rename(Bucket.t, BSON.ObjectId.t, String.t) :: Mongo.result(BSON.document)
+  def rename(%Bucket{topology_pid: topology_pid, opts: opts} = bucket, file_id, new_filename) do
+    query       = %{_id: file_id}
+    update      = %{"$set" => %{filename: new_filename}}
+    collection  = files_collection_name(bucket)
+    {:ok, _doc} = Mongo.find_one_and_update(topology_pid, collection, query, update, opts)
+  end
+
+  @doc """
+  Given a `id`, delete this stored file’s files collection document and
+  associated chunks from a GridFS bucket.
+  """
+  @spec delete(Bucket.t, String.t) :: Mongo.result(BSON.document)
+  def delete(%Bucket{} = bucket, file_id) when is_binary(file_id) do
+    delete(bucket, ObjectId.decode!(file_id))
+  end
+
+  @spec delete(Bucket.t, BSON.ObjectId.t) :: Mongo.result(BSON.document)
+  def delete(%Bucket{topology_pid: topology_pid, opts: opts} = bucket, %BSON.ObjectId{} = oid) do
+    # first delete files document
+    collection = files_collection_name(bucket)
+    {:ok, %Mongo.DeleteResult{deleted_count: _}} = Mongo.delete_one(topology_pid, collection, %{_id: oid}, opts)
+
+    # then delete all chunk documents
+    collection = chunks_collection_name(bucket)
+    {:ok, %Mongo.DeleteResult{deleted_count: _}} = Mongo.delete_many(topology_pid, collection, %{files_id: oid}, opts)
```

Comment:
Use `with` here to be able to capture the errors instead of having match errors should something go wrong.